StreamProcessorHealthTest#shouldMarkUnhealthyWhenReprocessingRetryLoop flaky

**Description**
Not reproducible locally on two different machines but often failing on the CI.


```

[INFO] No dependency problems found

[ERROR] Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 14.833 s <<< FAILURE! - in io.zeebe.engine.processor.StreamProcessorHealthTest

[ERROR] io.zeebe.engine.processor.StreamProcessorHealthTest.shouldMarkUnhealthyWhenReprocessingRetryLoop  Time elapsed: 10.786 s  <<< FAILURE!

org.junit.ComparisonFailure: expected:<[tru]e> but was:<[fals]e>

	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)

	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)

	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)

	at io.zeebe.test.util.TestUtil$Invocation.until(TestUtil.java:91)

	at io.zeebe.test.util.TestUtil$Invocation.until(TestUtil.java:64)

	at io.zeebe.test.util.TestUtil.waitUntil(TestUtil.java:36)

	at io.zeebe.engine.processor.StreamProcessorHealthTest.shouldMarkUnhealthyWhenReprocessingRetryLoop(StreamProcessorHealthTest.java:87)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

	at java.base/java.lang.reflect.Method.invoke(Method.java:566)

	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)

	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)

	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)

	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)

	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)

	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

	at org.junit.rules.RunRules.evaluate(RunRules.java:20)

	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)

	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)

	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)

	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)

	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)

	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)

	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)

	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)

	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)

	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)

	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)

	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)

	at org.junit.runners.Suite.runChild(Suite.java:128)

	at org.junit.runners.Suite.runChild(Suite.java:27)

	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)

	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)

	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)

	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)

	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)

	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)

	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)

	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)

	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)

	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)

	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)

	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)

	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)



[INFO] 

[INFO] Results:

[INFO] 

[ERROR] Failures: 

[ERROR]   StreamProcessorHealthTest.shouldMarkUnhealthyWhenReprocessingRetryLoop:87 expected:<[tru]e> but was:<[fals]e>

[INFO] 

[ERROR] Tests run: 1054, Failures: 1, Errors: 0, Skipped: 0
```

--- a/engine/pom.xml
+++ b/engine/pom.xml
@@ -140,13 +140,6 @@
       <artifactId>zeebe-test-util</artifactId>
       <scope>test</scope>
     </dependency>
-
-    <dependency>
-      <groupId>io.zeebe</groupId>
-      <artifactId>atomix-cluster</artifactId>
-      <scope>test</scope>
-    </dependency>
-
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-core</artifactId>

--- a/engine/src/test/java/io/zeebe/engine/processor/StreamProcessorHealthTest.java
+++ b/engine/src/test/java/io/zeebe/engine/processor/StreamProcessorHealthTest.java
@@ -12,6 +12,7 @@ import static io.zeebe.engine.processor.TypedRecordProcessors.processors;
 import static io.zeebe.protocol.record.intent.WorkflowInstanceIntent.ELEMENT_ACTIVATED;
 import static io.zeebe.protocol.record.intent.WorkflowInstanceIntent.ELEMENT_ACTIVATING;
 import static io.zeebe.test.util.TestUtil.waitUntil;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
 
 import io.zeebe.engine.state.ZeebeState;
@@ -23,10 +24,13 @@ import io.zeebe.protocol.record.RejectionType;
 import io.zeebe.protocol.record.ValueType;
 import io.zeebe.protocol.record.intent.Intent;
 import io.zeebe.util.health.HealthStatus;
+import io.zeebe.util.sched.Actor;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 import org.junit.After;
+import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 
@@ -36,17 +40,24 @@ public class StreamProcessorHealthTest {
 
   private StreamProcessor streamProcessor;
   private TypedStreamWriter mockedLogStreamWriter;
-  private final AtomicBoolean shouldFlushThrowException = new AtomicBoolean();
-  private final AtomicInteger invocation = new AtomicInteger();
-  private final AtomicBoolean shouldFailErrorHandlingInTransaction = new AtomicBoolean();
-  private final AtomicBoolean shouldProcessingThrowException = new AtomicBoolean(true);
+  private AtomicBoolean shouldFlushThrowException;
+  private AtomicInteger invocation;
+  private AtomicBoolean shouldFailErrorHandlingInTransaction;
+  private AtomicBoolean shouldProcessingThrowException;
+
+  @Before
+  public void before() {
+    invocation = new AtomicInteger();
+    shouldFlushThrowException = new AtomicBoolean();
+    shouldFailErrorHandlingInTransaction = new AtomicBoolean();
+    shouldProcessingThrowException = new AtomicBoolean();
+  }
 
   @After
-  public void after() {
-    // To exit the processing error loop
-    shouldProcessingThrowException.set(false);
+  public void tearDown() {
     shouldFlushThrowException.set(false);
     shouldFailErrorHandlingInTransaction.set(false);
+    shouldProcessingThrowException.set(false);
   }
 
   @Test
@@ -67,13 +78,23 @@ public class StreamProcessorHealthTest {
   @Test
   public void shouldMarkUnhealthyWhenReprocessingRetryLoop() {
     // given
+    shouldProcessingThrowException.set(true);
     final long firstPosition =
         streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
     streamProcessorRule.writeWorkflowInstanceEventWithSource(ELEMENT_ACTIVATED, 1, firstPosition);
+    assertThat(
+            streamProcessorRule
+                .events()
+                .onlyWorkflowInstanceRecords()
+                .withIntent(ELEMENT_ACTIVATED)
+                .exists())
+        .isTrue();
 
     streamProcessor = getErrorProneStreamProcessor();
+    final var healthStatusCheck = HealthStatusCheck.of(streamProcessor);
+    streamProcessorRule.getActorSchedulerRule().submitActor(healthStatusCheck);
 
-    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+    waitUntil(() -> healthStatusCheck.hasHealthStatus(HealthStatus.HEALTHY));
     waitUntil(() -> invocation.get() > 1);
 
     // when
@@ -84,21 +105,24 @@ public class StreamProcessorHealthTest {
     streamProcessorRule.getClock().addTime(HEALTH_CHECK_TICK_DURATION.multipliedBy(1));
 
     // then
-    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+    waitUntil(() -> healthStatusCheck.hasHealthStatus(HealthStatus.UNHEALTHY));
   }
 
   @Test
   public void shouldMarkUnhealthyWhenOnErrorHandlingWriteEventFails() {
     // given
     streamProcessor = getErrorProneStreamProcessor();
-    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+    final var healthStatusCheck = HealthStatusCheck.of(streamProcessor);
+    streamProcessorRule.getActorSchedulerRule().submitActor(healthStatusCheck);
+
+    waitUntil(() -> healthStatusCheck.hasHealthStatus(HealthStatus.HEALTHY));
 
     // when
     shouldFlushThrowException.set(true);
     streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
 
     // then
-    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+    waitUntil(() -> healthStatusCheck.hasHealthStatus(HealthStatus.UNHEALTHY));
   }
 
   @Test
@@ -119,15 +143,20 @@ public class StreamProcessorHealthTest {
   @Test
   public void shouldMarkUnhealthyWhenExceptionErrorHandlingInTransaction() {
     // given
+    shouldProcessingThrowException.set(true);
     streamProcessor = getErrorProneStreamProcessor();
-    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+    final var healthStatusCheck = HealthStatusCheck.of(streamProcessor);
+    streamProcessorRule.getActorSchedulerRule().submitActor(healthStatusCheck);
+    waitUntil(() -> healthStatusCheck.hasHealthStatus(HealthStatus.HEALTHY));
 
     // when
+    // since processing fails we will write error event
+    // we want to fail error even transaction
     shouldFailErrorHandlingInTransaction.set(true);
     streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
 
     // then
-    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+    waitUntil(() -> healthStatusCheck.hasHealthStatus(HealthStatus.UNHEALTHY));
   }
 
   @Test
@@ -266,6 +295,24 @@ public class StreamProcessorHealthTest {
         throw new RuntimeException("flush failed");
       }
       return wrappedWriter.flush();
+    }
+  }
+
+  private static final class HealthStatusCheck extends Actor {
+    private final StreamProcessor streamProcessor;
+
+    private HealthStatusCheck(final StreamProcessor streamProcessor) {
+      this.streamProcessor = streamProcessor;
+    }
+
+    public static HealthStatusCheck of(final StreamProcessor streamProcessor) {
+      return new HealthStatusCheck(streamProcessor);
+    }
+
+    public boolean hasHealthStatus(final HealthStatus healthStatus) {
+      return actor
+          .call(() -> streamProcessor.getHealthStatus() == healthStatus)
+          .join(5, TimeUnit.SECONDS);
     }
   }
 }

--- a/engine/src/test/java/io/zeebe/engine/processor/StreamProcessorInconsistentPositionTest.java
+++ b/engine/src/test/java/io/zeebe/engine/processor/StreamProcessorInconsistentPositionTest.java
@@ -17,10 +17,10 @@ import static io.zeebe.test.util.TestUtil.waitUntil;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
 
+import io.zeebe.engine.util.ListLogStorage;
 import io.zeebe.engine.util.RecordStream;
 import io.zeebe.engine.util.StreamProcessingComposite;
 import io.zeebe.engine.util.TestStreams;
-import io.zeebe.logstreams.util.AtomixLogStorageRule;
 import io.zeebe.protocol.record.ValueType;
 import io.zeebe.test.util.AutoCloseableRule;
 import io.zeebe.util.sched.clock.ControlledActorClock;
@@ -49,21 +49,13 @@ public final class StreamProcessorInconsistentPositionTest {
   private TestStreams testStreams;
 
   @Before
-  public void setup() throws Exception {
+  public void setup() {
 
     testStreams = new TestStreams(tempFolder, closeables, actorSchedulerRule.get());
 
-    final var dataDirectory = tempFolder.newFolder();
-
-    final AtomixLogStorageRule logStorageRule = new AtomixLogStorageRule(tempFolder, 1);
-    logStorageRule.open(
-        b ->
-            b.withDirectory(dataDirectory)
-                .withMaxEntrySize(4 * 1024 * 1024)
-                .withMaxSegmentSize(128 * 1024 * 1024));
-
-    testStreams.createLogStream(getLogName(1), 1, logStorageRule);
-    testStreams.createLogStream(getLogName(2), 2, logStorageRule);
+    final var listLogStorage = new ListLogStorage();
+    testStreams.createLogStream(getLogName(1), 1, listLogStorage);
+    testStreams.createLogStream(getLogName(2), 2, listLogStorage);
 
     firstStreamProcessorComposite =
         new StreamProcessingComposite(testStreams, 1, DEFAULT_DB_FACTORY);

--- a/engine/src/test/java/io/zeebe/engine/util/EngineRule.java
+++ b/engine/src/test/java/io/zeebe/engine/util/EngineRule.java
@@ -125,6 +125,7 @@ public final class EngineRule extends ExternalResource {
   protected void after() {
     subscriptionHandlerExecutor.shutdown();
     environmentRule = null;
+    subscriptionHandlers.clear();
   }
 
   public void start() {

--- a/engine/src/test/java/io/zeebe/engine/util/ListLogStorage.java
+++ b/engine/src/test/java/io/zeebe/engine/util/ListLogStorage.java
@@ -7,7 +7,6 @@
  */
 package io.zeebe.engine.util;
 
-import io.atomix.raft.zeebe.ZeebeEntry;
 import io.zeebe.logstreams.spi.LogStorage;
 import io.zeebe.logstreams.spi.LogStorageReader;
 import java.io.IOException;
@@ -22,7 +21,7 @@ import org.agrona.DirectBuffer;
 public class ListLogStorage implements LogStorage {
 
   private final ConcurrentNavigableMap<Long, Integer> positionIndexMapping;
-  private final List<ZeebeEntry> entries;
+  private final List<Entry> entries;
   private LongConsumer positionListener;
 
   public ListLogStorage() {
@@ -50,8 +49,8 @@ public class ListLogStorage implements LogStorage {
           return OP_RESULT_NO_DATA;
         }
 
-        final var zeebeEntry = entries.get(index);
-        final var data = zeebeEntry.data();
+        final var entry = entries.get(index);
+        final var data = entry.getData();
         readBuffer.wrap(data, data.position(), data.remaining());
 
         return address + 1;
@@ -89,16 +88,14 @@ public class ListLogStorage implements LogStorage {
       final ByteBuffer blockBuffer,
       final AppendListener listener) {
     try {
-      final var zeebeEntry =
-          new ZeebeEntry(
-              0, System.currentTimeMillis(), lowestPosition, highestPosition, blockBuffer);
-      entries.add(zeebeEntry);
+      final var entry = new Entry(lowestPosition, highestPosition, blockBuffer);
+      entries.add(entry);
       final var index = entries.size();
       positionIndexMapping.put(lowestPosition, index);
       listener.onWrite(index);
 
       if (positionListener != null) {
-        positionListener.accept(zeebeEntry.highestPosition());
+        positionListener.accept(entry.getHighestPosition());
       }
       listener.onCommit(index);
     } catch (final Exception e) {
@@ -126,4 +123,28 @@ public class ListLogStorage implements LogStorage {
 
   @Override
   public void flush() throws Exception {}
+
+  private static final class Entry {
+    private final long lowestPosition;
+    private final long highestPosition;
+    private final ByteBuffer data;
+
+    public Entry(final long lowestPosition, final long highestPosition, final ByteBuffer data) {
+      this.lowestPosition = lowestPosition;
+      this.highestPosition = highestPosition;
+      this.data = data;
+    }
+
+    public long getLowestPosition() {
+      return lowestPosition;
+    }
+
+    public long getHighestPosition() {
+      return highestPosition;
+    }
+
+    public ByteBuffer getData() {
+      return data;
+    }
+  }
 }

--- a/engine/src/test/java/io/zeebe/engine/util/ListLogStorage.java
+++ b/engine/src/test/java/io/zeebe/engine/util/ListLogStorage.java
@@ -1,0 +1,129 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.engine.util;
+
+import io.atomix.raft.zeebe.ZeebeEntry;
+import io.zeebe.logstreams.spi.LogStorage;
+import io.zeebe.logstreams.spi.LogStorageReader;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.LongConsumer;
+import org.agrona.DirectBuffer;
+
+public class ListLogStorage implements LogStorage {
+
+  private final ConcurrentNavigableMap<Long, Integer> positionIndexMapping;
+  private final List<ZeebeEntry> entries;
+  private LongConsumer positionListener;
+
+  public ListLogStorage() {
+    this.entries = new CopyOnWriteArrayList<>();
+    this.positionIndexMapping = new ConcurrentSkipListMap<>();
+  }
+
+  public void setPositionListener(final LongConsumer positionListener) {
+    this.positionListener = positionListener;
+  }
+
+  @Override
+  public LogStorageReader newReader() {
+    return new LogStorageReader() {
+      @Override
+      public boolean isEmpty() {
+        return entries.isEmpty();
+      }
+
+      @Override
+      public long read(final DirectBuffer readBuffer, final long address) {
+        final var index = (int) (address - 1);
+
+        if (index < 0 || index >= entries.size()) {
+          return OP_RESULT_NO_DATA;
+        }
+
+        final var zeebeEntry = entries.get(index);
+        final var data = zeebeEntry.data();
+        readBuffer.wrap(data, data.position(), data.remaining());
+
+        return address + 1;
+      }
+
+      @Override
+      public long readLastBlock(final DirectBuffer readBuffer) {
+        return read(readBuffer, entries.size());
+      }
+
+      @Override
+      public long lookUpApproximateAddress(final long position) {
+
+        if (position == Long.MIN_VALUE) {
+          return entries.isEmpty() ? OP_RESULT_INVALID_ADDR : 1;
+        }
+
+        final var lowerIndex = positionIndexMapping.lowerEntry(position);
+        if (lowerIndex != null) {
+          return lowerIndex.getValue();
+        }
+
+        return 1;
+      }
+
+      @Override
+      public void close() {}
+    };
+  }
+
+  @Override
+  public void append(
+      final long lowestPosition,
+      final long highestPosition,
+      final ByteBuffer blockBuffer,
+      final AppendListener listener) {
+    try {
+      final var zeebeEntry =
+          new ZeebeEntry(
+              0, System.currentTimeMillis(), lowestPosition, highestPosition, blockBuffer);
+      entries.add(zeebeEntry);
+      final var index = entries.size();
+      positionIndexMapping.put(lowestPosition, index);
+      listener.onWrite(index);
+
+      if (positionListener != null) {
+        positionListener.accept(zeebeEntry.highestPosition());
+      }
+      listener.onCommit(index);
+    } catch (final Exception e) {
+      listener.onWriteError(e);
+    }
+  }
+
+  @Override
+  public void open() throws IOException {}
+
+  @Override
+  public void close() {
+    entries.clear();
+  }
+
+  @Override
+  public boolean isOpen() {
+    return true;
+  }
+
+  @Override
+  public boolean isClosed() {
+    return false;
+  }
+
+  @Override
+  public void flush() throws Exception {}
+}

--- a/engine/src/test/java/io/zeebe/engine/util/StreamProcessorRule.java
+++ b/engine/src/test/java/io/zeebe/engine/util/StreamProcessorRule.java
@@ -103,6 +103,10 @@ public final class StreamProcessorRule implements TestRule {
             .around(new FailedTestRecordPrinter());
   }
 
+  public ActorSchedulerRule getActorSchedulerRule() {
+    return actorSchedulerRule;
+  }
+
   @Override
   public Statement apply(final Statement base, final Description description) {
     return chain.apply(base, description);

--- a/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
+++ b/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
@@ -52,6 +52,7 @@ import java.nio.file.Path;
 import java.time.Duration;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
@@ -232,7 +233,7 @@ public final class TestStreams {
             .onProcessedListener(mockOnProcessedListener)
             .streamProcessorFactory(factory)
             .build();
-    streamProcessor.openAsync().join();
+    streamProcessor.openAsync().join(15, TimeUnit.SECONDS);
 
     final var asyncSnapshotDirector =
         new AsyncSnapshotDirector(
@@ -241,7 +242,7 @@ public final class TestStreams {
             currentSnapshotController,
             stream.getAsyncLogStream(),
             snapshotInterval);
-    actorScheduler.submitActor(asyncSnapshotDirector).join();
+    actorScheduler.submitActor(asyncSnapshotDirector).join(15, TimeUnit.SECONDS);
 
     final LogContext context = logContextMap.get(logName);
     final ProcessorContext processorContext =

--- a/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
+++ b/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
@@ -28,9 +28,9 @@ import io.zeebe.logstreams.log.LogStreamBatchWriter;
 import io.zeebe.logstreams.log.LogStreamReader;
 import io.zeebe.logstreams.log.LogStreamRecordWriter;
 import io.zeebe.logstreams.log.LoggedEvent;
+import io.zeebe.logstreams.spi.LogStorage;
 import io.zeebe.logstreams.state.SnapshotStorage;
 import io.zeebe.logstreams.state.StateSnapshotController;
-import io.zeebe.logstreams.util.AtomixLogStorageRule;
 import io.zeebe.logstreams.util.SyncLogStream;
 import io.zeebe.logstreams.util.SynchronousLogStream;
 import io.zeebe.logstreams.util.TestSnapshotStorage;
@@ -44,7 +44,6 @@ import io.zeebe.protocol.record.intent.Intent;
 import io.zeebe.test.util.AutoCloseableRule;
 import io.zeebe.util.Loggers;
 import io.zeebe.util.sched.ActorScheduler;
-import java.io.File;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.nio.file.FileAlreadyExistsException;
@@ -112,40 +111,39 @@ public final class TestStreams {
   }
 
   public SynchronousLogStream createLogStream(final String name, final int partitionId) {
-    final File segments;
-    try {
-      segments = dataDirectory.newFolder(name, "segments");
-    } catch (final IOException e) {
-      throw new UncheckedIOException(e);
-    }
-
-    return createLogStream(name, partitionId, segments);
+    final var listLogStorage = new ListLogStorage();
+    return createLogStream(
+        name,
+        partitionId,
+        listLogStorage,
+        logStream -> listLogStorage.setPositionListener(logStream::setCommitPosition));
   }
 
-  public SyncLogStream createLogStream(
-      final String name, final int partitionId, final File segmentsDir) {
-    final AtomixLogStorageRule logStorageRule =
-        new AtomixLogStorageRule(dataDirectory, partitionId);
-    logStorageRule.open(
-        b ->
-            b.withDirectory(segmentsDir)
-                .withMaxEntrySize(4 * 1024 * 1024)
-                .withMaxSegmentSize(128 * 1024 * 1024));
-    return createLogStream(name, partitionId, logStorageRule);
+  public SynchronousLogStream createLogStream(
+      final String name, final int partitionId, final ListLogStorage listLogStorage) {
+    return createLogStream(
+        name,
+        partitionId,
+        listLogStorage,
+        logStream -> listLogStorage.setPositionListener(logStream::setCommitPosition));
   }
 
-  public SyncLogStream createLogStream(
-      final String name, final int partitionId, final AtomixLogStorageRule logStorageRule) {
+  public SynchronousLogStream createLogStream(
+      final String name,
+      final int partitionId,
+      final LogStorage logStorage,
+      final Consumer<SyncLogStream> logStreamConsumer) {
     final var logStream =
         SyncLogStream.builder()
             .withLogName(name)
-            .withLogStorage(logStorageRule.getStorage())
+            .withLogStorage(logStorage)
             .withPartitionId(partitionId)
             .withActorScheduler(actorScheduler)
             .build();
-    logStorageRule.setPositionListener(logStream::setCommitPosition);
 
-    final LogContext logContext = LogContext.createLogContext(logStream, logStorageRule);
+    logStreamConsumer.accept(logStream);
+
+    final LogContext logContext = LogContext.createLogContext(logStream, logStorage);
     logContextMap.put(name, logContext);
     closeables.manage(logContext);
     closeables.manage(() -> logContextMap.remove(name));
@@ -357,31 +355,30 @@ public final class TestStreams {
       writer.metadataWriter(metadata);
       writer.valueWriter(value);
 
-      return doRepeatedly(() -> writer.tryWrite()).until(p -> p >= 0);
+      return doRepeatedly(writer::tryWrite).until(p -> p >= 0);
     }
   }
 
   private static final class LogContext implements AutoCloseable {
     private final SynchronousLogStream logStream;
-    private final AtomixLogStorageRule logStorageRule;
+    private final LogStorage logStorage;
     private final LogStreamRecordWriter logStreamWriter;
 
-    private LogContext(
-        final SynchronousLogStream logStream, final AtomixLogStorageRule logStorageRule) {
+    private LogContext(final SynchronousLogStream logStream, final LogStorage logStorage) {
       this.logStream = logStream;
       logStreamWriter = logStream.newLogStreamRecordWriter();
-      this.logStorageRule = logStorageRule;
+      this.logStorage = logStorage;
     }
 
     public static LogContext createLogContext(
-        final SyncLogStream logStream, final AtomixLogStorageRule logStorageRule) {
-      return new LogContext(logStream, logStorageRule);
+        final SyncLogStream logStream, final LogStorage logStorage) {
+      return new LogContext(logStream, logStorage);
     }
 
     @Override
     public void close() {
       logStream.close();
-      logStorageRule.close();
+      logStorage.close();
     }
 
     public LogStreamRecordWriter getLogStreamWriter() {

